### PR TITLE
Add `SDDMM` example

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @hameerabbasi,

This PR adds `SDDMM` example and upgrades Finch to the latest version.

**[UPDATED 14.05.2024]**
For my machine, running:
```bash
python examples/sddmm_example.py
```
gives:
```
Finch
Took 8.787564675013224 s.

Numba
Took 22.904020706812542 s.

SciPy
Took 22.59452811876933 s.
```

### CODE REVIEW

**Comment by mtsokol** [comment]
I think density could be increased to `0.0001` so we have 100 non-zeros (more realistic?) - I get same performance.

File: examples/sddmm_example.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+DENSITY = 0.0001
+ITERS = 3
+rng = np.random.default_rng()
+
+
+def benchmark(func, info, args):
+    print(info)
+    start = time.time()
+    for i in range(ITERS):
+        print(f"iter: {i}")
```
Comment (hameerabbasi) [inline examples/sddmm_example.py:18]:
```suggestion
```

`print` statements have a large overhead

File: examples/sddmm_example.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        benchmark(sddmm_finch, info="Finch", args=[s, a, b])
+
+    # Numba
+    with sparse.Backend(backend=sparse.BackendType.Numba):
+        s = sparse.random((LEN, LEN), density=DENSITY, random_state=rng)
+        a = rng.random((LEN, LEN))
+        b = rng.random((LEN, LEN))
+
+        def sddmm_numba(s, a, b):
+            return sparse.sum(s[:, :, None] * (a[:, None, :] * b[None, :, :]), axis=-1)
```
Comment (hameerabbasi) [inline examples/sddmm_example.py:48]: We should go with what's fastest for each backend.
```suggestion
            return s * (a @ b)
```

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
Two final changes then this is ready.

**Comment by hameerabbasi** [comment]
I'd actually like to test the examples as well. Can we add something like the following to CI:

```bash
# test_examples.sh
for example in $(find ./examples/ -iname *.py); do
  python $example
done

# in CI
source test_examples.sh
```

**Review by hameerabbasi — DISMISSED** [review]

File: examples/sddmm_example.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        result_numba = sddmm_numba(s, a, b)
+        assert sparse.nonzero(result_numba)[0].size > 5
+        # Benchmark
+        benchmark(sddmm_numba, info="Numba", args=[s, a, b])
+
+    # SciPy
+    def sddmm_scipy(s, a, b):
+        return s * (a @ b)
+
+    s = sps.random_array((LEN, LEN), format="coo", density=DENSITY, random_state=rng)
```
Comment (hameerabbasi) [inline examples/sddmm_example.py:59]:
```suggestion
    s = sps.random_array((LEN, LEN), format="csr", density=DENSITY, random_state=rng)
```

CSR should be better for SciPy.

**Comment by mtsokol** [comment]
Blocked by https://github.com/willow-ahrens/Finch.jl/issues/534

File: examples/sddmm_example.py (hunk 3 above)
Comment (mtsokol) [reply examples/sddmm_example.py]: Done!

**Comment by mtsokol** [comment]
Here's a debug output for Finch lazy mode plan:

<details>

```julia
Executing:
:(function var"##compute#410"(prgm)
      begin
          V = (((((((((((((((((((prgm.children[1]).children[2]).children[2]).children[3]).children[1]).children[1]).children[1]).children[2]).children[1]).children[2]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[2]).tns.val::Tensor{SparseCOOLevel{2, Tuple{Int64, Int64}, Vector{Int64}, Tuple{PlusOneVector{Int32}, PlusOneVector{Int32}}, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}
          V_2 = ((((((((((((((((((((((((((((prgm.children[1]).children[2]).children[2]).children[3]).children[1]).children[1]).children[1]).children[2]).children[1]).children[3]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[2]).children[1]).children[2]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[2]).tns.val::Tensor{DenseLevel{Int64, DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}}
          V_3 = ((((((((((((((((((((((((((((prgm.children[1]).children[2]).children[2]).children[3]).children[1]).children[1]).children[1]).children[2]).children[1]).children[3]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[2]).children[1]).children[3]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[1]).children[2]).tns.val::Tensor{DenseLevel{Int64, DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}}
          A0 = V::Tensor{SparseCOOLevel{2, Tuple{Int64, Int64}, Vector{Int64}, Tuple{PlusOneVector{Int32}, PlusOneVector{Int32}}, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}
          A0_2 = Tensor(Dense(SparseDict(Element{0.0, Float64}())))::Tensor{DenseLevel{Int64, SparseLevel{Int64, Finch.DictTable{Int64, Int64, Vector{Int64}, Vector{Int64}, Vector{Int64}, Dict{Tuple{Int64, Int64}, Int64}}, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}}
          @finch mode = :fast begin
                  A0_2 .= 0.0
                  for i1 = _
                      for i0 = _
                          A0_2[i1, i0] = A0[i0, i1]
                      end
                  end
                  return A0_2
              end
          A2 = V_2::Tensor{DenseLevel{Int64, DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}}
          A4 = V_3::Tensor{DenseLevel{Int64, DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, PyArray{Float64, 1, true, true, Float64}}}}}
          A8 = Tensor(Dense(SparseDict(Element{0.0, Float64}())))::Tensor{DenseLevel{Int64, SparseLevel{Int64, Finch.DictTable{Int64, Int64, Vector{Int64}, Vector{Int64}, Vector{Int64}, Dict{Tuple{Int64, Int64}, Int64}}, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}}
          @finch mode = :fast begin
                  A8 .= 0.0
                  for i52 = _
                      for i51 = _
                          for i50 = _
                              A8[i50, i51] << + >>= (*)(A0_2[i50, i51], (*)(A2[1, i52], A4[1, i52]))
                          end
                      end
                  end
                  return A8
              end
          return (A8,)
      end
  end)
```

</details>

**Comment by willow-ahrens** [comment]
Let's keep working on this until we see a speedup from fusion. I believe a fusion-based speedup should be achievable here, so it's a good goal to work towards.

**Review by hameerabbasi — APPROVED** [review]

**Comment by mtsokol** [comment]
Right now we have precompilation of a few kernels. This causes a timeout of the first benchmark. Let me fix it.

**Comment by willow-ahrens** [comment]
Thanks @mtsokol!

**Pull request merged**